### PR TITLE
Feature/median module

### PR DESCRIPTION
**Pull request by iamnotstatic**
**Description:**
This PR introduces a Solidity smart contract for a Median Price Oracle. The contract is designed to act as a trusted price reference, handling price updates from authorized nodes and signers. It includes functionality for authorization, minimum quorum requirements, and price validation.

**Code Changes:**
- Added a `PriceData` struct to store timestamp and price data.
- Added an array `priceHistory` to keep a history of price updates.
- Implemented functions for authorizing and deauthorizing nodes and signers.
- Added functions for updating the minimum quorum and reading the last price and timestamp.
- Introduced an `update` function to handle price updates from authorized signers.
- Implemented a `recover` function to verify signatures.
- Added a `median` function to calculate the median of an array of values.
- Included a `sort` function for sorting an array of values.

Please review and provide feedback on the proposed Median Price Oracle contract. This contract can serve as a critical component in various decentralized applications for ensuring reliable price feeds.

If there are any suggestions, improvements, or concerns, please let us know, and we'll be happy to address them.

### CODE REVIEW

**Comment by AmadiMichael** [comment]
- the recover function is vulnerable to signature malleability as it does not check if the s value is in less than the order of the curve / 2. Because of this, if a node updates the price, a malicious user can subtract s from the order of the curve and flip v to 27 if it's 28 or vice versa and still get a valid ecdsa signature and then update the price to whatever they want it to be to sufficiently exploit the protocol. This is possible because points on a curve are symettrical over the x axis
example
```rs
// if v, r and s is a valid sig to get another valid sig, i can do
// n is order of the curve
let new_v = v == 27 ? 28 : 27;
let new_r = r;
let new_s = n - s mod n
// new_v, new_r, and new_s is also a valid sig
```
best practice is to only take sigs that the s value is less than the order of the curve / 2.

- the hash of the message signed is not calculated according to this [eip standard](https://eips.ethereum.org/EIPS/eip-191). it should be prefixed with `\xEthereum Signed Message:\n` then signed, this is how libraries like ethers rs and js sign messages too. also in the rust relayer client, i abi.encode it. With abi.encodepacked which was used here, since `timestamp` is of uint32, it'll return a different bytes value.
- `lastPrice` and `lastTimestamp` can be packed to in a slot by using `uint128` as their types. this will save some gas as they'll be accessed within the same transaction when called and only needs one sload. I see you casted lastPrice into a uint128 in the update function too.

i want to create a pr with the fixes.

**Review by MeisterJustice — APPROVED** [review]
LGTM!!!!

**Comment by iamnotstatic** [comment]
> * the recover function is vulnerable to signature malleability as it does not check if the s value is in less than the order of the curve / 2. Because of this, if a node updates the price, a malicious user can subtract s from the order of the curve and flip v to 27 if it's 28 or vice versa and still get a valid ecdsa signature and then update the price to whatever they want it to be to sufficiently exploit the protocol. This is possible because points on a curve are symettrical over the x axis
>   example
> 
> ```rust
> // if v, r and s is a valid sig to get another valid sig, i can do
> // n is order of the curve
> let new_v = v == 27 ? 28 : 27;
> let new_r = r;
> let new_s = n - s mod n
> // new_v, new_r, and new_s is also a valid sig
> ```
> 
> best practice is to only take sigs that the s value is less than the order of the curve / 2.
> 
> * the hash of the message signed is not calculated according to this [eip standard](https://eips.ethereum.org/EIPS/eip-191). it should be prefixed with `\xEthereum Signed Message:\n` then signed, this is how libraries like ethers rs and js sign messages too. also in the rust relayer client, i abi.encode it. With abi.encodepacked which was used here, since `timestamp` is of uint32, it'll return a different bytes value.
> * `lastPrice` and `lastTimestamp` can be packed to in a slot by using `uint128` as their types. this will save some gas as they'll be accessed within the same transaction when called and only needs one sload. I see you casted lastPrice into a uint128 in the update function too.
> 
> i want to create a pr with the fixes.

Great catch. I appreciate your diligence in reviewing the code and identifying potential vulnerabilities and optimizations. Your suggestions are indeed important for enhancing the security and efficiency of the contract.

I'm looking forward to your PR with these fixes and optimizations. Please feel free to reach out if you have any questions or need further clarification on the code.

File: contracts/median.sol [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        uint256 _minimumQuorum,
+        bytes32 _currencyPair
+    ) public initializer {
+        __Ownable_init_unchained(msg.sender);
+
+        minimumQuorum = _minimumQuorum;
+        currencyPair = _currencyPair;
+    }
+
+    modifier onlyAuthorizedNode() {
```
Comment (njokuScript) [inline contracts/median.sol]: we don't need this modifier. Nodes never call the contracts. Only relayer clients should call write methods in here

File: contracts/median.sol [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    function deauthorizeNode(address _nodeAddress) external onlyOwner {
+        if (authorizedNodes[_nodeAddress]) {
+            authorizedNodes[_nodeAddress] = false;
+            authorizedNodesCount--;
+        }
+
+        emit DeauthorizedNode(_nodeAddress);
+    }
+
+    function authorizeSigner(address _signerAddress) external onlyOwner {
```
Comment (njokuScript) [inline contracts/median.sol]: I'd call this `authorizeRelayer` instead

File: contracts/median.sol [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        return (lastTimestamp, lastPrice);
+    }
+
+    function update(
+        uint256[] calldata _prices,
+        uint32[] calldata _timestamps,
+        uint8[] calldata _v,
+        bytes32[] calldata _r,
+        bytes32[] calldata _s
+    ) external onlyAuthorizedSigner hasMinimumQuorum {
```
Comment (njokuScript) [inline contracts/median.sol]: `onlyAuthorizedRelayer`

**Review by njokuScript — CHANGES_REQUESTED** [review]

File: contracts/median.sol [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+    modifier onlyAuthorizedSigner() {
+        if (!authorizedSigners[msg.sender]) {
+            revert OnlyAuthorizedSigners();
+        }
+        _;
+    }
+
+    modifier hasMinimumQuorum() {
+        if (authorizedNodesCount < minimumQuorum) {
```
Comment (njokuScript) [inline contracts/median.sol]: we shouldn't be using a modifier to check for minimumQuorum.

minimumQuorum should be checked when you get the prices.

so inside the `update` function, we'll have:
```
        if (price.length < minimumQuorum) {
```

**Review by njokuScript — CHANGES_REQUESTED** [review]

File: contracts/median.sol (hunk 1 above)
Comment (iamnotstatic) [reply contracts/median.sol]: > we don't need this modifier. Nodes never call the contracts. Only relayer clients should call write methods in here

Yeah that makes sense

**Pull request merged**